lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/BulkPartitionTask.scala

```diff
@@ -0,0 +1,130 @@
+  */
+class BulkPartitionTask extends BasePartitionTask {
+  override protected def initializeInternal(ctx: TrainingContext,
+                                            shouldExecuteTraining: Boolean,
+                                            isEmptyPartition: Boolean): Unit = {
+    // For useSingleDataset mode, we need to add to bulk synchronization stops
+    if (ctx.useSingleDatasetMode && !isEmptyPartition) {
+      ctx.sharedState().incrementArrayProcessedSignal(log)
+      if (!shouldExecuteTraining) ctx.sharedState().incrementDataPrepDoneSignal(log)
+    }
+  }
+
+  protected def preparePartitionDataInternal(ctx: PartitionTaskContext,
+                                             inputRows: Iterator[Row]): PartitionDataState = {
+    // In useSingleDataset mode, we need to synchronize start of data loading
+    if (ctx.shouldExecuteTraining) {
+      // If in useSingleDatasetMode, we need to set a signal that helpers can start
+      if (ctx.trainingCtx.useSingleDatasetMode) {
+        ctx.sharedState.helperStartSignal.countDown()
+        log.info(s"Initiated helper start signal on task ${ctx.taskId}, partition ${ctx.partitionId}")
+      }
+    } else {
+      log.info(s"Waiting for helper start signal on task ${ctx.taskId}, partition ${ctx.partitionId}")
+      ctx.sharedState.helperStartSignal.await()
+    }
+
+    // Store the chunked partition data in local memory and then add it to the aggregated data
+    val aggregatedColumns = {
+      log.info(s"Reading data on task ${ctx.taskId}, partition ${ctx.partitionId}")
+      val prepAggregatedColumns: BaseChunkedColumns = getChunkedColumns(ctx, inputRows)
+      log.info(s"Merging data on task ${ctx.taskId}, partition ${ctx.partitionId}")
+      mergeChunksIntoAggregatedArrays(ctx, prepAggregatedColumns, isForValidation = false)
+    }
+    val aggregatedValidationColumns = ctx.trainingCtx.validationData.map { data =>
+      val prepAggregatedColumns: BaseChunkedColumns = getChunkedColumns(ctx, data.value.toIterator)
+      mergeChunksIntoAggregatedArrays(ctx, prepAggregatedColumns, isForValidation = true)
+    }
+    PartitionDataState(Option(aggregatedColumns), aggregatedValidationColumns)
+  }
+
+  protected def generateFinalDatasetInternal(ctx: PartitionTaskContext,
+                                             dataState: PartitionDataState,
+                                             forValidation: Boolean,
+                                             referenceDataset: Option[LightGBMDataset]): LightGBMDataset = {
+    val ac = if (forValidation) dataState.aggregatedValidationData.get
+             else dataState.aggregatedTrainingData.get
+    try {
+      val datasetInner: LightGBMDataset = ac.generateDataset(ctx, referenceDataset)
+      ctx.trainingCtx.columnParams.groupColumn.foreach(_ => datasetInner.addGroupColumn(ac.getGroups))
+      datasetInner.setFeatureNames(ctx.trainingCtx.featureNames, ac.getNumCols)
+      datasetInner
+    } finally {
+      ac.cleanup()
```

Comment:
likewise here

Reply:
I just moved this logic and did not modify it.  But I believe it does the correct thing.